### PR TITLE
[Suggestions are welcome] Docs v0.8

### PR DESCRIPTION
**Pull request by justheuristic**
Planned full documentation:

* Quickstart
   - [x] run ffn server from console
   - [x] play with remote experts
   - [x] Update TODOs after #81 is merged
* MNIST - LATER
   - more detailed info: setting up server, learning rates, expert grid
   - advanced: async. training and why its useful
* Custom Experts - LATER
   - extending ExpertBackend
     - show multiple inputs/outputs
   - running from python and from console
   - advanced: dropout
* API reference
   - [x] hivemind.client
   - [x] hivemind.server
   - [x] hivemind.dht
* [x] Contributing
* [x] Related work

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -53,9 +53,10 @@ def run(self):
     name='hivemind',
     version=version_string,
     cmdclass={'install': ProtoCompileInstall, 'develop': ProtoCompileDevelop},
-    description='',
-    long_description='',
-    author='Learning@home authors',
+    description='Decentralized deep learning framework in pytorch.',
+    long_description='Decentralized deep learning framework in pytorch. Built to train GPT-3-sized models on '
```
Comment (mryab) [inline setup.py:57]: Mentions of GPT-3 are likely to become obsolete at some point

File: hivemind/server/__init__.py [hunk 2]
```diff
@@ -113,8 +117,36 @@ def create(listen_on='0.0.0.0:*', num_experts=None, expert_uids=None, expert_cls
         # initialize experts
         if expert_uids is None:
             num_experts = num_experts if num_experts is not None else 1
-            expert_uids = [f'{expert_prefix}{hivemind.DHT.UID_DELIMITER}{i + expert_offset}'
-                           for i in range(num_experts)]
+            expert_pattern = expert_pattern if expert_pattern is not None else f"expert.[0:{num_experts}]"
+            print(expert_pattern)
```
Comment (mryab) [inline hivemind/server/__init__.py:121]: del

File: hivemind/server/__init__.py [hunk 3]
```diff
@@ -113,8 +117,36 @@ def create(listen_on='0.0.0.0:*', num_experts=None, expert_uids=None, expert_cls
         # initialize experts
         if expert_uids is None:
             num_experts = num_experts if num_experts is not None else 1
```
Comment (mryab) [inline hivemind/server/__init__.py:119]: If it's None, better if we fail with an error here

File: hivemind/server/__init__.py [hunk 4]
```diff
@@ -113,8 +117,36 @@ def create(listen_on='0.0.0.0:*', num_experts=None, expert_uids=None, expert_cls
...
+
+            # parse expert pattern
+            expert_pattern_blocks = expert_pattern.split(hivemind.DHT.UID_DELIMITER)
+
+            def generate_uid():
+                uid = ''
+                for block in expert_pattern_blocks:
+                    try:
+                        if '[' not in block and ']' not in block:
+                            uid += block
```
Comment (mryab) [inline hivemind/server/__init__.py:131]: String concatenation is not free, better to keep them in a list and use .join() afterwards

File: hivemind/server/__init__.py [hunk 5]
```diff
@@ -1,8 +1,10 @@
+from __future__ import annotations
 import multiprocessing as mp
 import multiprocessing.synchronize
 import threading
 from contextlib import contextmanager
 
+import numpy as np
```
Comment (mryab) [inline hivemind/server/__init__.py:7]: Do we really need numpy here just for randint generation?

File: hivemind/server/__init__.py [hunk 6]
```diff
@@ -62,33 +64,35 @@ def __init__(
...
+         e.g. ('123.11.22.33:1337', '[fe80::abe2:db1c:be7d:5a85]:4567'), default = no peers
         :param dht_port:  DHT node will listen on this port, default = find open port
         You can then use this node as initial peer for subsequent servers.
         :param verbose: whether to print server started / finished / terminated events
         :param start: if True, starts server right away and returns when server is ready for requests
         """
-        assert (expert_uids is None) != (num_experts is None and expert_prefix == 'expert' and expert_offset == 0), \
-            "Please provide either expert uids *or* (num_experts, expert_prefix and expert_offset), not both"
+        assert (expert_pattern is None) or (expert_uids is None), \
+            "Please provide either expert_uids *or* num_experts and expert_pattern, but not both"
```
Comment (mryab) [inline hivemind/server/__init__.py:93]: This function has too many responsibilities now; better to at least extract the creation of ExpertBackends in a separate function

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/server/__init__.py (hunk 4 above)
Comment (justheuristic) [reply hivemind/server/__init__.py:131]: thx, fixed

File: hivemind/server/__init__.py (hunk 6 above)
Comment (justheuristic) [reply hivemind/server/__init__.py:93]: extracted to a separate function

File: hivemind/server/__init__.py (hunk 5 above)
Comment (justheuristic) [reply hivemind/server/__init__.py:7]: switched to random

File: hivemind/server/__init__.py (hunk 2 above)
Comment (justheuristic) [reply hivemind/server/__init__.py]: thanks!

File: hivemind/server/__init__.py (hunk 3 above)
Comment (justheuristic) [reply hivemind/server/__init__.py:119]: agreed, eta 1h

**Review by mryab — APPROVED** [review]
